src/Dependencies/Collections/ImmutableSegmentedList`1.cs

```diff
@@ -0,0 +1,632 @@
+            builder.InsertRange(index, items);
+            return builder.ToImmutable();
+        }
+
+        public int LastIndexOf(T item, int index, int count, IEqualityComparer<T>? equalityComparer)
+        {
+            var self = this;
+
+            if (index < 0)
+                ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+            if (count < 0 || count > self.Count)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count);
+            if (index - count + 1 < 0)
+                throw new ArgumentException();
+
+            if (count == 0 && index == 0 && self.IsEmpty)
+            {
+                // SegmentedList<T> doesn't allow starting at index 0 for an empty list, but IImmutableList<T> does.
+                // Handle it explicitly to avoid an exception.
+                return -1;
+            }
+
+            return self._list.LastIndexOf(item, index, count, equalityComparer);
+        }
+
+        /// <inheritdoc cref="ImmutableList{T}.Remove(T)"/>
+        public ImmutableSegmentedList<T> Remove(T value)
+        {
+            var self = this;
+
+            var index = self.IndexOf(value);
+            if (index < 0)
+                return self;
+
+            return self.RemoveAt(index);
+        }
+
+        /// <inheritdoc cref="ImmutableList{T}.Remove(T, IEqualityComparer{T}?)"/>
+        public ImmutableSegmentedList<T> Remove(T value, IEqualityComparer<T>? equalityComparer)
+        {
+            var self = this;
+
+            var index = self.IndexOf(value, 0, Count, equalityComparer);
+            if (index < 0)
+                return self;
+
+            return self.RemoveAt(index);
+        }
+
+        /// <inheritdoc cref="ImmutableList{T}.RemoveAll(Predicate{T})"/>
+        public ImmutableSegmentedList<T> RemoveAll(Predicate<T> match)
+        {
+            // TODO: Optimize this to avoid allocations if no items are removed
+            // TODO: Optimize this to share pages prior to the first removed item
+            var builder = ToValueBuilder();
+            builder.RemoveAll(match);
+            return builder.ToImmutable();
+        }
+
+        /// <inheritdoc cref="ImmutableList{T}.RemoveAt(int)"/>
+        public ImmutableSegmentedList<T> RemoveAt(int index)
+        {
+            // TODO: Optimize this to share pages prior to the removed item
+            var builder = ToValueBuilder();
+            builder.RemoveAt(index);
+            return builder.ToImmutable();
+        }
+
+        /// <inheritdoc cref="ImmutableList{T}.RemoveRange(IEnumerable{T})"/>
+        public ImmutableSegmentedList<T> RemoveRange(IEnumerable<T> items)
+        {
+            if (items is null)
+                throw new ArgumentNullException(nameof(items));
+
+            var self = this;
+
+            if (self.IsEmpty)
+                return self;
+
+            var builder = ToValueBuilder();
+            foreach (var item in items)
```

Comment:
➡️ I was worried originally that this could be bad, but then I saw it's the [same thing](https://github.com/dotnet/runtime/blob/61109138bc1d9bf98088a7dfe439ef3abac566db/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList_1.cs#L326-L333) done by `ImmutableList<T>`.

Reply:
That doesn't necessarily mean it's the right way to do it. My benchmark is bad, (apparently because I can't call IndexOf on a SegmentedList?) but the sort -> walk through once made it scale almost optimally. It got 120x slower when I increased the size of the segmented list by 100x and for an input list of 10000 items, it only got 4% slower when the number of items to remove increased from 10 to 80, which sounds pretty good to me. (I ignored some pointer issues necessary to keep it unsorted, so it wouldn't be quite that good in practice.)

Reply:
We chatted offline, and it sounds like although this would be better for large data sets, that's not a common scenario, so we should leave this as-is.